tests/test_data_sources.py

```diff
@@ -0,0 +1,595 @@
+        columns=[
+            "company_id",
+            "company_employees",
+            "company_revenue",
+            "employee_id",
+            "employee_occupation",
+            "employee_age",
+            "employee_promotion",
+            "tags",
+        ],
+    )
+
+    return df
+
+
+@pytest.fixture
+def company_has_employee_multi_df():
+    df = pd.DataFrame(
+        [
+            (1, 100, 1000, 0, 0, 25, 0, [1, 2, 3]),
+            (1, 100, 1000, 1, 1, 35, 1, [1, 2]),
+            (1, 100, 1000, 3, 3, 45, 0, [3, 4]),
+            (2, 5000, 100000, 4, 1, 18, 1, [1, 4]),
+            (2, 5000, 100000, 5, 1, 20, 1, [1, 1]),
+            (2, 5000, 100000, 6, 4, 31, 0, [1, 2]),
+        ],
+        columns=[
+            "company_id",
+            "company_employees",
+            "company_revenue",
+            "employee_id",
+            "employee_occupation",
+            "employee_age",
+            "employee_promotion",
+            "tags",
+        ],
+    )
+
+    company_df = df[
+        ["company_id", "company_employees", "company_revenue"]
+    ].drop_duplicates()
+    employee_df = df[
+        ["employee_id", "employee_occupation", "employee_age", "employee_promotion"]
+    ].drop_duplicates()
+    tags_df = df.explode("tags")[["tags"]].drop_duplicates().astype(int)
+    company_has_employee_df = df[["company_id", "employee_id"]].drop_duplicates()
+    company_has_tag_df = (
+        df[["company_id", "tags"]].explode("tags").drop_duplicates().astype(int)
+    )
+
+    return company_df, employee_df, tags_df, company_has_employee_df, company_has_tag_df
+
+
+def test_inconsistent_graph_specs_missing_tag_specs(company_has_employee_df):
+    # given
+    company_node_params = structures.NodeTypeParams(
+        node_type_name="company",
+        id_col="company_id",
+        attributes=("company_employees", "company_revenue"),
+        attr_type="float",
+    )
+
+    employee_node_params = structures.NodeTypeParams(
+        node_type_name="employee",
+        id_col="employee_id",
+        attributes=("employee_occupation", "employee_age"),
+        label_col="employee_promotion",
+        attr_type="long",
+    )
+
+    company_has_emp_edge_params = structures.EdgeTypeParams(
+        edge_type_name="has",
+        source_name="company",
+        target_name="employee",
+        source_id_col="company_id",
+        target_id_col="employee_id",
+    )
+
+    company_has_tag_edge_params = structures.EdgeTypeParams(
+        edge_type_name="has",
+        source_name="company",
+        target_name="tag",
+        source_id_col="company_id",
+        target_id_col="tags",
+        multivalue_target=True,
+    )
+
+    df_source_specs = data_sources.DataFrameSpecs(
+        name="df1",
+        node_params=(
+            company_node_params,
+            employee_node_params,
+        ),
+        edge_params=(company_has_emp_edge_params, company_has_tag_edge_params),
+        data_frame=company_has_employee_df,
+    )
+
```

Comment:
trivial: If you added `#given` then add also `# when # then` (or something like that) to be consistent 

```python
def test_inconsistent_graph_specs_missing_tag_specs(company_has_employee_df):
    # given
    company_node_params = structures.NodeTypeParams(
        node_type_name="company",
        id_col="company_id",
        attributes=("company_employees", "company_revenue"),
        attr_type="float",
    )
...
  
    # when # then
    with pytest.raises(ValueError) as e:
        data_sources.GraphSpecs(data_sources=(df_source_specs,))
```

Reply:
Or even better add some assertion for errors

```python
        # when
        with pytest.raises(ValueError) as excinfo:
            data_sources.GraphSpecs(data_sources=(df_source_specs,))

        # then
        assert (
            str(excinfo.value)
            == "Some error message"
        )